File: backend/app/agents/web_sources.py
"""Normalize web-search tool results into UI-friendly source records."""
from __future__ import annotations

from typing import Any

from ..schemas.chat import WebSource
from ..tools.types import ToolResult

WEB_SOURCE_TOOLS = {
    "tavily_search",
    "tavily_extract",
    "tavily_research",
}


def _text(value: Any) -> str:
    return value.strip() if isinstance(value, str) else ""


def _score(value: Any) -> float | None:
    if isinstance(value, (int, float)):
        return float(value)
    return None


def _source_from_item(item: Any, *, source_tool: str) -> WebSource | None:
    if not isinstance(item, dict):
        return None
    url = _text(item.get("url"))
    if not url:
        return None
    title = _text(item.get("title")) or url
    preview = (
        _text(item.get("content"))
        or _text(item.get("raw_content"))
        or _text(item.get("text"))
    )
    return WebSource(
        title=title,
        url=url,
        content_preview=preview or None,
        score=_score(item.get("score")),
        source_tool=source_tool,
    )


def extract_web_sources(result: ToolResult) -> list[WebSource]:
    if result.tool not in WEB_SOURCE_TOOLS:
        return []

    items: list[Any] = []
    raw_results = result.data.get("results")
    if isinstance(raw_results, list):
        items.extend(raw_results)
    elif isinstance(raw_results, dict):
        nested = raw_results.get("results")
        if isinstance(nested, list):
            items.extend(nested)

    if result.tool == "tavily_extract" and not items:
        items.append(result.data)

    sources: list[WebSource] = []
    seen_urls: set[str] = set()
    for item in items:
        source = _source_from_item(item, source_tool=result.tool)
        if source is None or source.url in seen_urls:
            continue
        sources.append(source)
        seen_urls.add(source.url)
    return sources

File: frontend/lib/webSources.ts
import type { ToolResult, WebSource } from "./types";

const WEB_SOURCE_TOOLS = new Set([
  "tavily_search",
  "tavily_extract",
  "tavily_research",
]);

export function webSourcesFromToolResult(result: ToolResult): WebSource[] {
  if (!WEB_SOURCE_TOOLS.has(result.tool)) return [];
  const rawResults = result.data?.results;
  const items = Array.isArray(rawResults)
    ? rawResults
    : typeof result.data?.url === "string"
      ? [result.data]
      : [];
  const seen = new Set<string>();
  const sources: WebSource[] = [];
  for (const raw of items) {
    if (!raw || typeof raw !== "object") continue;
    const item = raw as Record<string, unknown>;
    const url = typeof item.url === "string" ? item.url : "";
    if (!url || seen.has(url)) continue;
    sources.push({
      title: typeof item.title === "string" && item.title ? item.title : url,
      url,
      content_preview:
        typeof item.content === "string"
          ? item.content
          : typeof item.raw_content === "string"
            ? item.raw_content
            : null,
      score: typeof item.score === "number" ? item.score : null,
      source_tool: result.tool,
    });
    seen.add(url);
  }
  return sources;
}

export function mergeWebSources(
  current: WebSource[] | null | undefined,
  incoming: WebSource[]
): WebSource[] {
  const merged = [...(current ?? [])];
  const seen = new Set(merged.map((source) => source.url));
  for (const source of incoming) {
    if (seen.has(source.url)) continue;
    merged.push(source);
    seen.add(source.url);
  }
  return merged;
}
